README


This archive contains the distribution Text-VimColor,
version 0.27:

  Syntax highlight text using Vim

This software is copyright (c) 2002-2006 by Geoff Richards.

This software is copyright (c) 2011 by Randy Stauner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v6.010.



Changes
Revision history for Text-VimColor

0.27      2017-08-07T01:42:02Z

  - Fix installer for 5.24.1+

0.26      2015-03-21T04:14:25Z

  - Fix test failures for vim 6.
  - Use Test::File::ShareDir to simplify testing.

0.25      2015-02-28T23:41:39Z

  - Add 'all_syntax_groups' option to enable additional syntax groups
    instead of just the primary ones.  Thanks Hinrik Örn Sigurðsson!

0.24      2014-04-13T18:16:06Z

  - Simplify encoding tests and mark a few as TODO
    while the library undergoes some refactoring.

0.23      2013-02-02T16:28:57Z

  - Attempt to do the right thing with character strings:
    Encode them in UTF-8, tell vim the file encoding (UTF-8),
    and return a (decoded) character string.
    Thanks to Ricardo Signes for the very helpful report (gh-1).

0.22      2012-10-29T23:36:29Z

  - The included 'text-vimcolor' script will attempt to load
    Win32::Console::ANSI on windows when printing ANSI to STDOUT
    in an attempt to make the output more useful.
  - Add a basic test to ensure the script still operates

0.21      2012-09-07T00:02:48Z

  - Add 'extra_vim_options' to be able to append options to the defaults

0.20      2012-08-11T02:05:18Z

  - Releasing as stable.

0.19-TRIAL 2012-07-28T03:44:54Z

  - Use IPC::Open3 for portability instead of doing fork/exec manually.
  - Various other fixes to attempt to make this usable on Windows.
    Thanks to Yasuhiro Matsumoto (mattn) for helping to port to Windows (gh-5).

0.18      2012-07-21T02:22:55Z

  [Bug Fix]
  - bin/text-vimcolor: Exit after printing usage when --help is specified
    (previous version continued to wait for STDIN).
    Thanks to Vyacheslav Matyukhin for the pull req (gh-6).

  [Installaion]
  - Try to show any available version information before aborting Makefile.PL

0.17      2012-03-22T02:49:27Z

  - Minor installation and test fixes for older systems

0.16      2012-03-21T17:38:45Z

  [Installation]
  - Simplify version detection using vim script

  [Testing]
  - Mark encoding tests as TODO tests for now (still under development)

0.15      2012-03-18T23:43:36Z

  [Performance]
  - Remove extraneous vim statement that caused delays on
    some files.  Thanks to Joe Mullally (jhackworth)
    for finding this and for the vim profiling tips (gh-2).

  [Installation]
  - Abort Makefile.PL before writing Makefile if minimum vim not found
  - Try not to hang the installation if an old vim is found
  - Lower Term::ANSIColor requirement to 1.03
    and make bright colors normal if Term::ANSIColor version < 3

  [Documentation]
  - Document that the minimum vim version is 6.
    (It has been that way since 2003.)

  [Testing]
  - Allow debugging based on TEXT_VIMCOLOR_DEBUG env var
  - Attempt to make tests more portable across vim versions
    (addressing rt-75829 and a few sparse smoke reports).
  - Reorganize the test suite increasing coverage and number.

0.14      2011-10-20T17:09:53Z

  - Add test to confirm fix of rt-50646.
  - Silence 'untie' warnings.

0.13      2011-08-15T23:10:00Z

  [Prereqs]
  - Exit Makefile.PL early if `vim` fails to execute.
    (Nothing else will work after that.)
  - Bump required version of Term::ANSIColor to 3.0.
    (Please let me know if this causes problems for anyone.)

  [Dist Meta Changes]
  - Convert to Dist::Zilla for releasing
  - Use File::ShareDir to manage included files consistently
  - Simplify and organize documentation and tests

0.12     2011-08-09T19:33:16Z

  [Enhancements]
  - Add Term::ANSIColor support
    Closes RT #69338.
  - Add '-n' (no swapfile) option (thanks hinrik.sig).
    Closes RT #45623.

  [Bug Fixes]
  - Set $ENV{HOME} to keep user-customized syntax files from
    breaking tests.  Closes RT #68513.
  - Untie STD(IN|OUT|ERR) in child process (thanks shiar).
    Closes RT #50646.

  [Maintenance Status Change]
  - Co-maintenance granted to Randy Stauner (RWSTAUNER) by PAUSE admins.
    Much appreciation to Geoff Richards (GEOFFR) for his initial work.

0.11     2006-02-19T19:32:27Z

  * No code changes, just packaging fixes.

0.10     2006-02-19T00:22:35Z

  * Be more careful about how 'exec' is called (belt & braces)

  * Improve error reporting.  Errors printed by the 'vim' command,
    as well as problems reported from 'exec' (like not being able
    to find the program) will now be reported (by getting passed
    through a temporary file).  Also, some 'die' calls were changed
    to 'croak'.

  * Allow the 'filetype' option to be overridden for each file
    processed, so that an object can be reused for multiple files
    of different types.

  * Provide a help/usage message in the text-vimcolor program.

  * Document which other CPAN modules use Text::VimColor, since
    you might find one of them be more useful than using it
    directly.

  * Various small corrections to the documentation.

0.09     2005-02-19T13:28:28Z

  * Allow Vim options to be set with ':let' and provide a default
    option to make shell scripts be interpretted as Bash syntax,
    since it's a superset of Bourne shell.  Added corresponding
    options to 'text-vimcolor' script.

  * Added "+set nomodeline" option to Vim for belt-and-braces
    security.  Vim shouldn't read modelines with the default
    configuration enforced by the other options, but there's
    no harm being extra cautious.  Suggested by Ian Langworth.

  * If the 'vim' command fails, print the correct exit code in
    the error message.  Thanks to Smylers for the fix, and for
    reporting all the following bugs.
    (closes CPAN bug #11556).

  * Add documentation for the 'filetype' option to 'new()'.
    (closes CPAN bug #11558).

  * Document and test for using a reference to a scalar rather
    than just a plain scalar as the 'string' argument to 'new()'
    and the argument to 'syntax_mark_string()'.  It was already
    implemented but was mistakenly documented as a TODO.
    (closes CPAN bug #11557).

  * Document potential problem from CPAN bug #11555.


t/lib/TVC_Test.pm
# vim: set ts=2 sts=2 sw=2 expandtab smarttab:
#
# This file is part of Text-VimColor
#
# This software is copyright (c) 2011 by Randy Stauner.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict;
use warnings;

package # hide from indexer
  TVC_Test;

# don't allow user-customized syntax files to throw off test results
$ENV{HOME} = 't';

use Text::VimColor;
use Path::Class qw(file dir);
use Exporter ();
our @ISA = qw(Exporter);
our @EXPORT = qw(
  file
  dir
  slurp_data
);

sub slurp_data {
  my ($filename) = @_;
  $filename = file('t', 'data', $filename)->stringify;
  open my $file, '<', $filename
    or die "error opening file '$filename': $!";

  return do { local $/; <$file> };
}

1;


t/data/has_tabs.pl
#
# This file is part of Text-VimColor
#
# This software is copyright (c) 2011 by Randy Stauner.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
# Perl script containing tabs.

if (@ARGV) {
	for (1 .. shift) {
		print "Number $_",
		      "\n";
	}
}


inc/VimCommand.pm
#
# This file is part of Text-VimColor
#
# This software is copyright (c) 2011 by Randy Stauner.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict;
use warnings;

# This module is overly defensive, but hopefully will handle all portability issues
package inc::VimCommand;
#our $MIN = '5.4';
our $MIN = '6.0';

use File::Temp qw( tempfile ); # core
use IO::File; # core
use IPC::Open3 'open3'; # core

# vim -h works at least as far back as 5.0 (--version not until 5.2)
# first line: "VIM - Vi IMproved 5.0 (date)"
our $VERSION_RE = qr/vim .+ ([0-9.]+)/i;

our $INFO = {};

our $MESSAGE = <<MESSAGE;
This module requires vim version $MIN or later
which must be compiled with at least the 'normal' feature set
(--with-features=normal) particularly including +syntax (obviously).

If you believe this to be an error please submit a bug report
including the output of `vim --version` if possible.
MESSAGE

sub import {
  require_minimum_with_message();
}

sub require_minimum_with_message {
  eval {
    require_minimum();
  };
  my $msg = $@;
  if( $msg ){
    $msg .= "\n$MESSAGE\n";
    $msg =~ s/^/# /mg;
    die $msg;
  }
}

sub require_minimum {
  my $info = info_from_script();
  my $ver = sprintf("%0.2f", $info->{version}/100);

  if( !$ver ){
    die "Unable to identify vim version ($ver)\n";
  }
  elsif( $ver < $MIN ){
    $ver =~ s/\.0/./; # 5.08 => 5.8
    die "Vim version $ver too low.\n";
  }
  elsif( !$info->{syntax} ){
    die "Vim does not have the +syntax feature.\n";
  }
  return $ver;
}

sub info_from_script {
  # touch output file
  my $out = write_temp('tvc-out-XXXXXX', '');

  # NOTE: use single quotes, or backslashed double quotes in the '=' expr:
  my $script = write_temp('tvc-script-XXXXXX', <<SCRIPT);
:put ='vim:' . version . ',syn:' . has('syntax') . ','
:0d
:write! $out
:quit!
SCRIPT

  # try
  eval {
    # use IPC::Open3 to prevent STDOUT/ERR from interfering with make/test
    vim(qw(-u NONE -s), $script);

    my $output = do { local $/; IO::File->new($out, 'r')->getline; };

    die "Vim script failed: Output file is empty.\n"
      unless $output;
    die "Failed to parse vim output:\n>$output\n"
      unless $output =~ /^vim:(\d+),syn:(.+?),$/;

    # if there was more info to get we could parse the output more generically
    $INFO = { version => $1, syntax => $2 };
  };
  # catch part 1
  my $e = $@;
  # finally
  unlink($out, $script);
  # catch part 2
  die $e if $e;

  return $INFO;
}

sub vim {
  my @args = @_;
  my $output;
  my $timeout = 10;

  # if we pass an arg vim doesn't understand (like --version before v5.2)
  # it will treat it like a filename and wait for input
  # so try not to hang indefinitely if we have an old version.
  eval {
    local $SIG{PIPE} = 'IGNORE';
    local $SIG{ALRM} = sub { die "alarm\n" };

    alarm $timeout;

    my $pid = open3(my ($i, $o), undef, vim => @args);

    # read handle before waitpid to avoid hanging (and timing out) on older systems
    $output = do { local $/; <$o>; };

    waitpid($pid, 0);
    #my $stat = $?;

    alarm 0;

    # open3 will probably die if vim isn't found (in unix environments)
    # we can't really trust the exit status to mean anything...
    # vim might exit 1 or 2 (or 0) for -h/--version, cmd.exe exits 1 if not found...
  };
  if( my $e = $@ ){
    $e = "Command aborted after $timeout seconds."
      if $e eq "alarm\n";
    die "Error attemting to execute 'vim':\n  $e\n";
  }
  return $output;
}

sub write_temp {
  my ($template, $text) = @_;
  my ($fh, $path) = tempfile( $template, TMPDIR => 1 );
  print $fh $text;
  close $fh;
  return $path;
}

1;


Makefile.PL
#
# This file is part of Text-VimColor
#
# This software is copyright (c) 2011 by Randy Stauner.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.010.
use strict;
use warnings;

use 5.006;

use lib ".";
use inc::VimCommand;
use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Syntax highlight text using Vim",
  "AUTHOR" => "Geoff Richards <qef\@laxan.com>, Randy Stauner <rwstauner\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06",
    "File::Temp" => 0,
    "IO::File" => 0,
    "IPC::Open3" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "DISTNAME" => "Text-VimColor",
  "EXE_FILES" => [
    "bin/text-vimcolor"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Text::VimColor",
  "PREREQ_PM" => {
    "Carp" => 0,
    "File::Copy" => 0,
    "File::ShareDir" => 0,
    "File::Temp" => 0,
    "Getopt::Long" => 0,
    "IO::File" => 0,
    "IPC::Open3" => 0,
    "Path::Class" => "0.04",
    "Symbol" => 0,
    "Term::ANSIColor" => "1.03",
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Path::Class" => "0.04",
    "Test::File::ShareDir::Dist" => 0,
    "Test::More" => "0.88",
    "blib" => "1.01",
    "lib" => 0
  },
  "VERSION" => "0.27",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Copy" => 0,
  "File::ShareDir" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "Getopt::Long" => 0,
  "IO::File" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Path::Class" => "0.04",
  "Symbol" => 0,
  "Term::ANSIColor" => "1.03",
  "Test::File::ShareDir::Dist" => 0,
  "Test::More" => "0.88",
  "blib" => "1.01",
  "constant" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O eq 'MSWin32' ) {
	$WriteMakefileArgs{PREREQ_PM}{'Win32::Console::ANSI'} = '0';
}

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
